### PR TITLE
Wrap more driver API calls

### PR DESCRIPTION
**Pull request by zasdfgbnm**

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
!build

**Comment by naoyam** [comment]
What happens when some functions are used through the dlsym interface, while some are used as usual. Would it potentially end up using two different libcuda.so?

**Comment by zasdfgbnm** [comment]
> What happens when some functions are used through the dlsym interface, while some are used as usual. Would it potentially end up using two different libcuda.so?

Depending on how the system/cmake is configured, I think this is possible.

**Comment by naoyam** [comment]
> > What happens when some functions are used through the dlsym interface, while some are used as usual. Would it potentially end up using two different libcuda.so?
> 
> Depending on how the system/cmake is configured, I think this is possible.

Isn't it potentially dangerous? It never seems like a good idea. 

What's the plan for addressing this issue? Are we going to wrap all libcuda functions that we need?

**Comment by zasdfgbnm** [comment]
> > > What happens when some functions are used through the dlsym interface, while some are used as usual. Would it potentially end up using two different libcuda.so?
> > 
> > 
> > Depending on how the system/cmake is configured, I think this is possible.
> 
> Isn't it potentially dangerous? It never seems like a good idea.
> 
> What's the plan for addressing this issue? Are we going to wrap all libcuda functions that we need?

I think we have no choice but to wrap all of them, because we are not supposed to directly link against libcuda.so. This PR should include all of them, but there's no way for me to check that...

File: csrc/driver_api.h [hunk 1]
```diff
@@ -22,7 +22,18 @@ namespace nvfuser {
 // List of driver APIs that you want the magic to happen.
 DECLARE_DRIVER_API_WRAPPER(cuGetErrorName);
 DECLARE_DRIVER_API_WRAPPER(cuGetErrorString);
+DECLARE_DRIVER_API_WRAPPER(cuModuleLoadDataEx);
```
Comment (jjsjann123) [inline csrc/driver_api.h:25]: nitpick on macro. `DECLARE_DRIVER_API_WRAPER` and `DEFINE_DRIVER_API_WRAPPER` seems to always show up in pairs. Maybe we can put them together as a macro and call wrapper on top?

something like
```
#define ALL_DRIVER_API_WRAPPER(fn) \
  fn(cuGetErrorName)                                  \
  fn(...)

// .h
ALL_DRIVER_API_WRAPPER(DECLARE_DRIVER_API_WRAPPER)
// .cpp
ALL_DRIVER_API_WRAPPER(DEFINE_DRIVER_API_WRAPPER)
```

**Review by jjsjann123 — APPROVED** [review]

**Comment by naoyam** [comment]
> > > > What happens when some functions are used through the dlsym interface, while some are used as usual. Would it potentially end up using two different libcuda.so?
> > > 
> > > 
> > > Depending on how the system/cmake is configured, I think this is possible.
> > 
> > 
> > Isn't it potentially dangerous? It never seems like a good idea.
> > What's the plan for addressing this issue? Are we going to wrap all libcuda functions that we need?
> 
> I think we have no choice but to wrap all of them, because we are not supposed to directly link against libcuda.so. This PR should include all of them, but there's no way for me to check that...

Could we have some simple CI check to detect accidental usage of the drive API? For example, we could flag all symbols whose name start with `cu` (but not `cuda`). We could replace all the valid dysym-based uses with `nvfuser::cu...` and exclude them from the check. This way, any future use of the driver API should be resolved to the ones in the `nvfuser` namespace.
@xwang233, what do you think?

**Comment by xwang233** [comment]
sounds reasonable to me, will implement a check for that

**Comment by nWEIdia** [comment]
FYI: https://gitlab-master.nvidia.com/dl/dgx/pytorch/-/jobs/70026691#L5223 building this branch and then "import nvfuser" would still encounter https://gitlab-master.nvidia.com/dl/dgx/pytorch/-/jobs/70026691#L5234 libcuda.so.1 not found issue.

File: csrc/driver_api.h (hunk 1 above)
Comment (zasdfgbnm) [reply csrc/driver_api.h:25]: Unfortunately you can not use `#if (CUDA_VERSION >= 12000)` inside `ALL_DRIVER_API_WRAPPER`...

**Comment by zasdfgbnm** [comment]
I am not sure if I have wrapped all driver API calls, and I don't know how to force cmake not to link `libcuda.so`. But I will merge this PR as is, because the remaining tasks belongs to a separate PR.

File: csrc/driver_api.h (hunk 1 above)
Comment (jjsjann123) [reply csrc/driver_api.h:25]: We can put the #if outside.

Something like these: https://github.com/pytorch/pytorch/blob/a565f1bee68478c0237836d5e593601e9194083b/aten/src/ATen/cuda/nvrtc_stub/ATenNVRTC.h#L62-L70

Comment (zasdfgbnm) [reply csrc/driver_api.h:25]: Good point! Let's include this into whatever next PR is for this topic.

**Comment by zasdfgbnm** [comment]
> > > > > What happens when some functions are used through the dlsym interface, while some are used as usual. Would it potentially end up using two different libcuda.so?
> > > > 
> > > > 
> > > > Depending on how the system/cmake is configured, I think this is possible.
> > > 
> > > 
> > > Isn't it potentially dangerous? It never seems like a good idea.
> > > What's the plan for addressing this issue? Are we going to wrap all libcuda functions that we need?
> > 
> > 
> > I think we have no choice but to wrap all of them, because we are not supposed to directly link against libcuda.so. This PR should include all of them, but there's no way for me to check that...
> 
> Could we have some simple CI check to detect accidental usage of the drive API? For example, we could flag all symbols whose name start with `cu` (but not `cuda`). We could replace all the valid dysym-based uses with `nvfuser::cu...` and exclude them from the check. This way, any future use of the driver API should be resolved to the ones in the `nvfuser` namespace. @xwang233, what do you think?

With https://github.com/NVIDIA/Fuser/issues/940, we can just check `ldd build/libnvfuser_codegen.so` and assert that there is no `libcuda.so`.

**Comment by xwang233** [comment]
I did a search and here are some results on `cuXXX` string matching. Does any file need to be excluded?

```
root@349daaeed493:/opt/pytorch/nvfuser# grep -IrP '^[^/]*\bcu[A-Z]' --exclude '*driver_api*' benchmark/ csrc/ examples/ nvfuser/ test/
csrc/cuda_utils.h:      cuGetErrorName(_result, &name);  \
csrc/cuda_utils.h:      cuGetErrorString(_result, &msg); \
csrc/tma.cpp:  NVFUSER_CUDA_SAFE_CALL(cuTensorMapEncodeTiled(
csrc/executor.cpp:  NVFUSER_CUDA_SAFE_CALL(cuOccupancyMaxActiveBlocksPerMultiprocessor(
csrc/executor.cpp:    NVFUSER_CUDA_SAFE_CALL(cuFuncGetAttribute(
csrc/executor.cpp:    NVFUSER_CUDA_SAFE_CALL(cuFuncGetAttribute(
csrc/executor.cpp:    NVFUSER_CUDA_SAFE_CALL(cuFuncSetAttribute(
csrc/executor.cpp:      NVFUSER_CUDA_SAFE_CALL(cuOccupancyMaxActiveBlocksPerMultiprocessor(
csrc/executor.cpp:      FUSER_PERF_SCOPE("ExecutorRunFusion::cuLaunchKernel");
csrc/executor.cpp:      NVFUSER_CUDA_SAFE_CALL(cuLaunchKernel(
csrc/executor.cpp:      FUSER_PERF_SCOPE("ExecutorRunFusion::cuLaunchCooperativeKernel");
csrc/executor.cpp:      NVFUSER_CUDA_SAFE_CALL(cuLaunchCooperativeKernel(
csrc/executor.cpp:  NVFUSER_CUDA_SAFE_CALL(cuLaunchKernel(
csrc/executor_utils.cpp:    NVFUSER_CUDA_SAFE_CALL(cuModuleLoadDataEx(
csrc/executor_utils.cpp:  NVFUSER_CUDA_SAFE_CALL(cuModuleGetFunction(
csrc/executor_utils.cpp:  NVFUSER_CUDA_SAFE_CALL(cuModuleGetFunction(
csrc/kernel_cache.cpp:    NVFUSER_CUDA_SAFE_CALL(cuDeviceGetAttribute(
csrc/kernel_cache.cpp:    NVFUSER_CUDA_SAFE_CALL(cuDeviceGetAttribute(
csrc/kernel_cache.cpp:        cuDeviceGetName(gpuname.data(), 100, args.getDeviceIndex()));
nvfuser/include/nvfuser/cuda_utils.h:      cuGetErrorName(_result, &name);  \
nvfuser/include/nvfuser/cuda_utils.h:      cuGetErrorString(_result, &msg); \
```

**Comment by jjsjann123** [comment]
I think anything inside our code base is implicitly using `nvfuser::cuXXX`.
We can make that explicit as Naoya suggested earlier.

**Comment by zasdfgbnm** [comment]
> I did a search and here are some results on `cuXXX` string matching. Does any file need to be excluded?
> 
> ```
> root@349daaeed493:/opt/pytorch/nvfuser# grep -IrP '^[^/]*\bcu[A-Z]' --exclude '*driver_api*' benchmark/ csrc/ examples/ nvfuser/ test/
> csrc/cuda_utils.h:      cuGetErrorName(_result, &name);  \
> csrc/cuda_utils.h:      cuGetErrorString(_result, &msg); \
> csrc/tma.cpp:  NVFUSER_CUDA_SAFE_CALL(cuTensorMapEncodeTiled(
> csrc/executor.cpp:  NVFUSER_CUDA_SAFE_CALL(cuOccupancyMaxActiveBlocksPerMultiprocessor(
> csrc/executor.cpp:    NVFUSER_CUDA_SAFE_CALL(cuFuncGetAttribute(
> csrc/executor.cpp:    NVFUSER_CUDA_SAFE_CALL(cuFuncGetAttribute(
> csrc/executor.cpp:    NVFUSER_CUDA_SAFE_CALL(cuFuncSetAttribute(
> csrc/executor.cpp:      NVFUSER_CUDA_SAFE_CALL(cuOccupancyMaxActiveBlocksPerMultiprocessor(
> csrc/executor.cpp:      FUSER_PERF_SCOPE("ExecutorRunFusion::cuLaunchKernel");
> csrc/executor.cpp:      NVFUSER_CUDA_SAFE_CALL(cuLaunchKernel(
> csrc/executor.cpp:      FUSER_PERF_SCOPE("ExecutorRunFusion::cuLaunchCooperativeKernel");
> csrc/executor.cpp:      NVFUSER_CUDA_SAFE_CALL(cuLaunchCooperativeKernel(
> csrc/executor.cpp:  NVFUSER_CUDA_SAFE_CALL(cuLaunchKernel(
> csrc/executor_utils.cpp:    NVFUSER_CUDA_SAFE_CALL(cuModuleLoadDataEx(
> csrc/executor_utils.cpp:  NVFUSER_CUDA_SAFE_CALL(cuModuleGetFunction(
> csrc/executor_utils.cpp:  NVFUSER_CUDA_SAFE_CALL(cuModuleGetFunction(
> csrc/kernel_cache.cpp:    NVFUSER_CUDA_SAFE_CALL(cuDeviceGetAttribute(
> csrc/kernel_cache.cpp:    NVFUSER_CUDA_SAFE_CALL(cuDeviceGetAttribute(
> csrc/kernel_cache.cpp:        cuDeviceGetName(gpuname.data(), 100, args.getDeviceIndex()));
> nvfuser/include/nvfuser/cuda_utils.h:      cuGetErrorName(_result, &name);  \
> nvfuser/include/nvfuser/cuda_utils.h:      cuGetErrorString(_result, &msg); \
> ```

I think as long as we are checking we are not linking with `libcuda.so`, we are fine? I don't think we additionally needs any other check?

**Comment by naoyam** [comment]
> > I did a search and here are some results on `cuXXX` string matching. Does any file need to be excluded?
> > ```
> > root@349daaeed493:/opt/pytorch/nvfuser# grep -IrP '^[^/]*\bcu[A-Z]' --exclude '*driver_api*' benchmark/ csrc/ examples/ nvfuser/ test/
> > csrc/cuda_utils.h:      cuGetErrorName(_result, &name);  \
> > csrc/cuda_utils.h:      cuGetErrorString(_result, &msg); \
> > csrc/tma.cpp:  NVFUSER_CUDA_SAFE_CALL(cuTensorMapEncodeTiled(
> > csrc/executor.cpp:  NVFUSER_CUDA_SAFE_CALL(cuOccupancyMaxActiveBlocksPerMultiprocessor(
> > csrc/executor.cpp:    NVFUSER_CUDA_SAFE_CALL(cuFuncGetAttribute(
> > csrc/executor.cpp:    NVFUSER_CUDA_SAFE_CALL(cuFuncGetAttribute(
> > csrc/executor.cpp:    NVFUSER_CUDA_SAFE_CALL(cuFuncSetAttribute(
> > csrc/executor.cpp:      NVFUSER_CUDA_SAFE_CALL(cuOccupancyMaxActiveBlocksPerMultiprocessor(
> > csrc/executor.cpp:      FUSER_PERF_SCOPE("ExecutorRunFusion::cuLaunchKernel");
> > csrc/executor.cpp:      NVFUSER_CUDA_SAFE_CALL(cuLaunchKernel(
> > csrc/executor.cpp:      FUSER_PERF_SCOPE("ExecutorRunFusion::cuLaunchCooperativeKernel");
> > csrc/executor.cpp:      NVFUSER_CUDA_SAFE_CALL(cuLaunchCooperativeKernel(
> > csrc/executor.cpp:  NVFUSER_CUDA_SAFE_CALL(cuLaunchKernel(
> > csrc/executor_utils.cpp:    NVFUSER_CUDA_SAFE_CALL(cuModuleLoadDataEx(
> > csrc/executor_utils.cpp:  NVFUSER_CUDA_SAFE_CALL(cuModuleGetFunction(
> > csrc/executor_utils.cpp:  NVFUSER_CUDA_SAFE_CALL(cuModuleGetFunction(
> > csrc/kernel_cache.cpp:    NVFUSER_CUDA_SAFE_CALL(cuDeviceGetAttribute(
> > csrc/kernel_cache.cpp:    NVFUSER_CUDA_SAFE_CALL(cuDeviceGetAttribute(
> > csrc/kernel_cache.cpp:        cuDeviceGetName(gpuname.data(), 100, args.getDeviceIndex()));
> > nvfuser/include/nvfuser/cuda_utils.h:      cuGetErrorName(_result, &name);  \
> > nvfuser/include/nvfuser/cuda_utils.h:      cuGetErrorString(_result, &msg); \
> > ```
> 
> I think as long as we are checking we are not linking with `libcuda.so`, we are fine? I don't think we additionally needs any other check?

Yes, just checking the `ldd` output should be fine. So, something like `if ldd libnvfuser_codegen.so | grep libcuda.so; then echo "Invalid linking"; fi`

**Comment by xwang233** [comment]
>Yes, just checking the ldd output should be fine. So, something like if ldd libnvfuser_codegen.so | grep libcuda.so; then echo "Invalid linking"; fi

This check was added to CI as "sanity check".

**Comment by naoyam** [comment]
Thanks @xwang233!

**Pull request merged**